src/librustc/lint/mod.rs

```diff
@@ -437,14 +437,27 @@ pub fn struct_lint_level<'a>(
     level: Level,
     src: LintSource,
     span: Option<MultiSpan>,
-    msg: &str,
+    msg: impl std::fmt::Display,
+) -> DiagnosticBuilder<'a> {
+    struct_lint_level_(sess, lint, level, src, span, &msg)
+}
+
+fn struct_lint_level_<'a>(
+    sess: &'a Session,
+    lint: &'static Lint,
+    level: Level,
+    src: LintSource,
+    span: Option<MultiSpan>,
+    msg: &dyn std::fmt::Display,
```

Comment:
Hmm... this does nothing though to prevent similar mistakes in the future.

Perhaps we should:

- Set the initial message to `""` (so there should be no allocation).

- Invert control: Instead of returning a `DiagnosticBuilder<'_>`, we take in a function `impl Fn(&mut DiagnosticBuilder<'_>)` which allows the caller to make modifications, but ultimately this is never called on `Level::Allow`, thereby avoiding allocations that follow in e.g. https://github.com/rust-lang/rust/blob/055ce62a1dfcd1f8d7b6dc224bf14af09ede39de/src/librustc_lint/builtin.rs#L190-L200

Reply:
I think the idea of changing the api surface to always defer evaluation of the message would be a relatively painless big win.

Reply:
Filed https://github.com/rust-lang/rust/issues/67927.